pandas/tests/test_frame.py

```diff
@@ -12385,6 +12404,93 @@ def verify(df):
                 for col in ['4th', '5th']:
                     verify(udf[col])
 
+        # GH7403
+        df = pd.DataFrame({'A': list('aaaabbbb'),'B':range(8), 'C':range(8)})
+        df.iloc[3, 1] = np.NaN
+        left = df.set_index(['A', 'B']).unstack(0)
+
+        vals = [[3, 0, 1, 2, nan, nan, nan, nan],
+                [nan, nan, nan, nan, 4, 5, 6, 7]]
+        vals = list(map(list, zip(*vals)))
+        idx = Index([nan, 0, 1, 2, 4, 5, 6, 7], name='B')
```

Comment:
On a related note, to see in general how messed up things are when have ``NaN``s in indices, consider the following:
```
In [698]: idx = MultiIndex.from_tuples([(1, 'a'), (np.nan, 'b')])

In [699]: idx.levels[0]
Out[699]: Int64Index([1], dtype='int64')

In [700]: idx.get_values()
Out[700]: array([(1.0, 'a'), (nan, 'b')], dtype=object)

In [701]: idx[:1].get_values()
Out[701]: array([(1, 'a')], dtype=object)
```
Why does ``idx.get_values()`` convert the int ``1`` to a float ``1.0`` just because there's a ``NaN`` somewhere else? (I can guess what the implementation is that would lead to this...) Is this a bug?